src/reusedBrowser.ts

```diff
@@ -160,17 +165,38 @@ export class ReusedBrowser implements vscodeTypes.Disposable {
       this._backend.on('stateChanged', params => {
         this._pageCountChanged(params.pageCount);
       });
-      this._backend.on('sourceChanged', params => {
-        if (!this._editor)
-          return;
-        const start = new this._vscode.Position(0, 0);
-        const end = new this._vscode.Position(Number.MAX_VALUE, Number.MAX_VALUE);
-        const range = this._editor.document.validateRange(new this._vscode.Range(start, end));
-        this._editor?.edit(async editBuilder => {
-          editBuilder.replace(range, params.text);
-          if (this._editor) {
-            const lastLine = this._editor.document.lineCount - 1;
-            this._editor.selections = [new this._vscode.Selection(lastLine, 0, lastLine, 0)];
+      this._backend.on('sourceChanged', async params => {
+        this._scheduleEdit(async () => {
+          if (!this._editor)
+            return;
+          if (this._isInlineEdit) {
+            if (params.actions.length > this._insertedEditActionCount) {
+              // Collapse selection to start recoding a new action.
+              this._editor.selections = [new this._vscode.Selection(this._editor.selection.end, this._editor.selection.end)];
+            }
+            this._insertedEditActionCount = params.actions.length;
+            if (params.actions?.length) {
+              await this._editor.edit(async editBuilder => {
+                if (!this._editor)
+                  return;
+                const indentation = guessIndentation(this._editor);
+                const action = params.actions[params.actions.length - 1].trim();
+                const lineNumber = this._editor.selection.start.line;
+                const selection = new this._vscode.Selection(new this._vscode.Position(lineNumber, 0), this._editor.selection.end);
+                editBuilder.replace(selection, indentation + action + '\n');
```

Comment:
How does this work with multiline actions?

Reply:
It does not, in a sense that they'll have poor formatting. I can follow up when I am able to add tests.